client/db/types.go

```diff
@@ -212,10 +212,51 @@ func (m *MetaMatch) ID() []byte {
 	return hashKey(append(m.Match.MatchID[:], m.Match.OrderID[:]...))
 }
 
+// Revoke marks this match as revoked and retires the match if no further
+// action is required. Recovery actions may be required for a revoked match
+// if the client has broadcasted a swap contract and has not redeemed the
+// counter-party's contract or refunded own swap.
+// Returns whether or not the match is retired by this revocation action.
+func (m *MetaMatch) Revoke(fromServer bool) bool {
+	m.MetaData.Proof.serverRevoked = fromServer
+	m.MetaData.Proof.selfRevoked = !fromServer
+
+	// Retire this match if no further action (refund or auto-redeem)
+	// is necessary. This should not be applied to cancel order matches.
+	wasRetired := m.MetaData.Retired
+	status, side := m.Match.Status, m.Match.Side
+	m.MetaData.Retired = m.Match.Address == "" || // cancel order match
+		status == order.NewlyMatched || // no swap sent yet, no recovery required
+		(status == order.MakerSwapCast && side == order.Taker) || // we're taker, haven't sent swap, no recovery required
+		(side == order.Maker && status == order.MakerRedeemed) || // we're maker and we've redeemed, no recovery required
+		status == order.MatchComplete // we've redeemed, no further action is necessary
+
+	return !wasRetired && m.MetaData.Retired
```

Comment:
Docs for this method should be clearer that it could return `false` if the match was **already** retired.

Reply:
> to avoid repeated revoke notifications

Yeah... DB updates may still need to be performed regardless.